Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -15,21 +15,71 @@ vector<uint> dockingModuleEquipmentIds;
 // Above how much cargo capacity, should a ship be rejected as a docking user?
 int cargoCapacityLimit = 275;
 
+// How much time will player be given before kick if carrier wants to jettison him.
+int jettisonKickTime = 15;
+
 uint connSystemID = CreateID("li06");
 
-///////////////////////////////////////////////////////////////////////////////////////////////////////////////
-// Function handling changing systems when undocking from a carrier
-void JumpToLocation(uint client, uint system, Vector pos, Matrix ornt)
+// Delayed actions, which need to be done. Look at HkTimerCheckKick().
+vector<ActionJettison> jettisonList;
+
+void HkTimerCheckKick()
 {
-	mapDeferredJumps[client].system = system;
-	mapDeferredJumps[client].pos = pos;
-	mapDeferredJumps[client].rot = ornt;
+	returncode = DEFAULT_RETURNCODE;
+
+	for (vector<ActionJettison>::iterator it = jettisonList.begin(); it != jettisonList.end();)
+	{
+		it->timeLeft--;
+		if (it->timeLeft == 0)
+		{
+			uint checkCarrierClientID = HkGetClientIdFromCharname(it->carrierCharname);
+			uint checkDockedClientID = HkGetClientIdFromCharname(it->dockedCharname);
+
+			// If both carrier and docked ship are still at server.
+			if (checkDockedClientID != -1 && checkCarrierClientID != -1)
+			{
+				// If the client is still docked
+				if (mobiledockClients[checkDockedClientID].wscDockedWithCharname == it->carrierCharname)
+				{
+					HkKickReason(ARG_CLIENTID(checkDockedClientID), L"Forced undocking.");
 
-	// Send the jump command to the client. The client will send a system switch out complete
-	// event which we intercept to set the new starting positions.
-	PrintUserCmdText(client, L" ChangeSys %u", system);
+					// If carrier died, we don't need this message.
+					if(!mobiledockClients[checkDockedClientID].carrierDied)
+						PrintUserCmdText(checkCarrierClientID, L"Jettisoning ship...");
+				}
+			}
+
+			jettisonList.erase(it);
+		}
+		else
+		{
+			it++;
+		}
+	}
+}
+
+void JettisonShip(uint carrierClientID, uint dockedClientID)
+{
+	const wchar_t* carrierCharname = (const wchar_t*)Players.GetActiveCharacterName(carrierClientID);
+	const wchar_t* dockedCharname = (const wchar_t*)Players.GetActiveCharacterName(dockedClientID);
+
+	if (dockedClientID != -1)
+	{
+		if (!mobiledockClients[dockedClientID].carrierDied)
+		{
+			PrintUserCmdText(carrierClientID, L"Ship warned. If it doesn't undock in %i seconds, it will be kicked by force.", jettisonKickTime);
+			PrintUserCmdText(dockedClientID, L"Carrier wants to jettison your ship. Undock willingly or you will be kicked after %i seconds.", jettisonKickTime);
+		}
+		else
+		{
+			PrintUserCmdText(dockedClientID, L"ALERT! Carrier ship is being destroyed. Launch ship or you will die in explosion.");
+		}
+		pub::Audio::PlaySoundEffect(dockedClientID, CreateID("rtc_klaxon_loop"));

```

Comment:
I don't think we should hardcode stuff like this. I get it was like this in the old plugin, but the old plugin was made with enough bad design decisions as it stands. I'd suggest setting it up as a config value that can be changed.

Reply:
Nice idea. Will be done in full version of the plugin.